### ISSUE TITLE
wc object not updating width and height if using mask

### ISSUE THREAD

**Issue by Xunius** [amueller/word_cloud#205]
Hi amueller,
Thanks for the hard work.
I was playing around with mask and re-coloring, and noticed that if I want to resize my coloring image to the same size as the generated text image, I had to query the mask's size rather than the wc object, e.g.

```
wc = WordCloud(background_color="white", max_words=2000, mask=alice_mask,
               stopwords=stopwords,width=width,height=height,random_state=100)
wc.generate(text)
newshape=wc.mask.shape
```
As stated in the API doc, `width` and `height` are ignored in the constructor if `mask` is given, but if I supply them, they will be stored in the `wc` obj. So I wonder will it be better to update these 2 properties in the constructor?

**Comment by amueller** [comment]
Sorry, I'm not sure what you mean by "query the mask size rather than the wc object". You mean that ``wc.width`` and ``wc.height`` are invalid? Yeah, that's true. We could set them in the constructor. I didn't really see a reason to do that, but I guess there's no harm.
The reliable way to get the shape of ``wc`` *after* you generated it is ``wc.to_array().shape``.

**Comment by Xunius** [comment]
Yes I meant `wc.width` and `wc.height` should be set to the size of the mask array, if a mask is given. Otherwise they will be, for instance (600,400) if they are given so in the constructor, even though the mask is actually (900,900). The reason is obvious: for the sake of consistency and the ease of use, the user doesn't have to read the API to figure out what they should query, but instead, the `width` and `height` names should be intuitive enough to infer these are the properties they want. Are they meant to infer the array size, or something else? what's the point of setting these 2 attributes in the first place?

**Comment by amueller** [comment]
Well they are meant to give the image size if you don't use a mask.
We could not set them if a mask is given or set them to the mask size. Feel free to submit a PR.

**Comment by richardfat7** [comment]
I found a bug related to this issue.

In wordcloud.py function generate_from_frequencies line [447](https://github.com/amueller/word_cloud/blob/b6f48e108224f84b0b1659cea8558c86ccfc9898/wordcloud/wordcloud.py#L447)
when it try to figure out the max font size from drawing first 2 words, it set max font size temporary as `self.height`, while it will work normally if we use `width` and `height` to specify the size, it could not work when only `mask` is supplied. The `self.height` will remain as the default 200 without respecting to the mask size which is the one used in drawing. If largest font size at the end is less than 200, then this will not be a problem, maybe that's why it is not easy to discover.

Code to reproduce the issue:
```
from wordcloud import WordCloud
import numpy as np

image_array = np.zeros((1200, 1600), dtype=np.uint32)
mask = image_array
q = {
    "Generating": 500,
    "WordCloud": 1000
}

# With Height and Width
wc = WordCloud(height=1200, width=1600, scale=0.25, background_color='white')
wc.generate_from_frequencies(q)
print(wc.layout_)

# With Mask only
wc = WordCloud(mask=mask, scale=0.25, background_color='white')
wc.generate_from_frequencies(q)
print(wc.layout_)

# With Mask and Height
wc = WordCloud(height=1200, mask=mask, scale=0.25, background_color='white')
wc.generate_from_frequencies(q)
print(wc.layout_)
```

Output:
```
# With Height and Width
[(('WordCloud', 1.0), 252, (250, 18), None, 'rgb(31, 161, 135)'), (('Generating', 0.5), 189, (882, 363), None, 'rgb(58, 186, 118)')]
# With Mask only
[(('WordCloud', 1.0), 171, (846, 524), None, 'rgb(40, 125, 142)'), (('Generating', 0.5), 128, (473, 178), None, 'rgb(42, 176, 127)')]
# With Mask and Height
[(('WordCloud', 1.0), 252, (961, 51), None, 'rgb(173, 220, 48)'), (('Generating', 0.5), 189, (707, 432), None, 'rgb(181, 222, 43)')]

**Comment by amueller** [comment]
Thanks, that looks like an issue. Do you want to send a PR to fix it? I'm a bit swamped right now.

**Comment by richardfat7** [comment]
I can help with the PR, but which direction you want me to go, temporarily override the self.height/self.width or change the self.height/width to the mask height and width from the initialization?

**Comment by amueller** [comment]
I think changing to the mask height/width from the initialization would be good. Thanks!

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — the fix discussed here looks like it landed.

Evidence:
- `WordCloud.__init__` now does exactly what @amueller proposed in this thread ("changing to the mask height/width from the initialization"):
  ```python
  # Override the width and height if there is a mask
  if mask is not None:
      self.width = mask.shape[1]
      self.height = mask.shape[0]
  ```
  ([wordcloud/wordcloud.py#L371-L374](https://github.com/amueller/word_cloud/blob/93c5e4cd50c5f2c56ec7eb624a620790dab5df6c/wordcloud/wordcloud.py#L371-L374))
- The class docstring also reflects the new contract: *"If mask is not None, width and height will be ignored and the shape of mask will be used instead."* ([L177-L182](https://github.com/amueller/word_cloud/blob/93c5e4cd50c5f2c56ec7eb624a620790dab5df6c/wordcloud/wordcloud.py#L177-L182))

So `wc.width` / `wc.height` are now consistent with the mask shape as soon as the constructor returns; no need to fall back to `wc.to_array().shape` for that.

Note: @richardfat7's nested observation about `max_font_size` being clamped to the default `self.height=200` during the two-word heuristic on line 447 is a separate (and narrower) bug. If that one is still reproducible on current `main` it would probably warrant its own dedicated issue rather than living under this title.

If I'm wrong on either count, point me at the doc/version and I'll dig further. Otherwise would you mind closing this out?

---
*Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against current `main`.*
